README.rst
========
emaillib
========

A simple library to ease the sending of emails with utf8 and attachments.


* Documentation: https://emaillib.readthedocs.org/en/latest


Development Workflow
====================

The workflow supports the following steps

 * lint
 * test
 * build
 * document
 * upload
 * graph

These actions are supported out of the box by the corresponding scripts under _CI/scripts directory with sane defaults based on best practices.
Sourcing setup_aliases.ps1 for windows powershell or setup_aliases.sh in bash on Mac or Linux will provide with handy aliases for the shell of all those commands prepended with an underscore.

The bootstrap script creates a .venv directory inside the project directory hosting the virtual environment. It uses pipenv for that.
It is called by all other scripts before they do anything. So one could simple start by calling _lint and that would set up everything before it tried to actually lint the project

Once the code is ready to be delivered the _tag script should be called accepting one of three arguments, patch, minor, major following the semantic versioning scheme.
So for the initial delivery one would call

    $ _tag --minor

which would bump the version of the project to 0.1.0 tag it in git and do a push and also ask for the change and automagically update HISTORY.rst with the version and the change provided.


So the full workflow after git is initialized is:

 * repeat as necessary (of course it could be test - code - lint :) )

   * code
   * lint
   * test
 * commit and push
 * develop more through the code-lint-test cycle
 * tag (with the appropriate argument)
 * build
 * upload (if you want to host your package in pypi)
 * document (of course this could be run at any point)


Important Information
=====================

This template is based on pipenv. In order to be compatible with requirements.txt so the actual created package can be used by any part of the existing python ecosystem some hacks were needed.
So when building a package out of this **do not** simple call

    $ python setup.py sdist bdist_egg

**as this will produce an unusable artifact with files missing.**
Instead use the provided build and upload scripts that create all the necessary files in the artifact.



Project Features
================

* Handles subject and message body as utf-8 by default
* Handles html and plain text
* Handles attachments either as list or comma delimited value arguments
* Supports CC and BCC
* Simply validates the provided recipient addresses


HISTORY.rst
.. :changelog:

History
-------

0.0.1 (16-09-2017)
---------------------

* First code creation


1.0.0 (07-07-2021)
------------------

* Dropped support for python2.7, updated the template and setup a pipeline for release.


tests/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: __init__.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html
"""

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''16-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


tests/test_emaillib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: test_emaillib.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Tests for `emaillib` module.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

from betamax.fixtures import unittest
from emaillib.emaillib import Message, SmtpServer, EasySender

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''16-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


class TestEmaillib(unittest.BetamaxTestCase):

    def setUp(self):
        """
        Test set up

        This is where you can setup things that you use throughout the tests. This method is called before every test.
        """
        pass

    def testNoSender(self):
        info = {'sender': '',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8'}
        with self.assertRaises(ValueError):
            message = Message(**info)

    def testInvalidSender(self):
        invalid = ['a',
                   'a@',
                   '@.',
                   '@a.',
                   '.@',
                   'a@@',
                   'a@@.']
        for sender in invalid:
            info = {'sender': sender,
                    'recipients': 'whatever@gmail.com',
                    'subject': 'Τεστ test',
                    'body': 'This is a τεστ on utf8'}
            with self.assertRaises(ValueError):
                message = Message(**info)

    def testNoRecipient(self):
        info = {'sender': 'whatever@gmail.com',
                'recipients': '',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8'}
        with self.assertRaises(ValueError):
            message = Message(**info)

    def testInvalidRecipient(self):
        invalid = ['a',
                   'a@',
                   '@.',
                   '@a.',
                   '.@',
                   'a@@',
                   'a@@.',
                   'a@a.com,a.',
                   'inv.domain.com,valid@test.com']
        for recipients in invalid:
            info = {'sender': 'whatever@gmail.com',
                    'recipients': recipients,
                    'subject': 'Τεστ test',
                    'body': 'This is a τεστ on utf8'}
            with self.assertRaises(ValueError):
                print(recipients)
                message = Message(**info)

    def testInvalidCC(self):
        invalid = ['a',
                   'a@',
                   '@.',
                   '@a.',
                   '.@',
                   'a@@',
                   'a@@.',
                   'a@a.com,a.',
                   'inv.domain.com,valid@test.com']
        for cc in invalid:
            info = {'sender': 'whatever@gmail.com',
                    'recipients': 'another@valid.com',
                    'cc': cc,
                    'subject': 'Τεστ test',
                    'body': 'This is a τεστ on utf8'}
            with self.assertRaises(ValueError):
                message = Message(**info)

    def testInvalidBCC(self):
        invalid = ['a',
                   'a@',
                   '@.',
                   '@a.',
                   '.@',
                   'a@@',
                   'a@@.',
                   'a@a.com,a.',
                   'inv.domain.com,valid@test.com']
        for bcc in invalid:
            info = {'sender': 'whatever@gmail.com',
                    'recipients': 'another@valid.com',
                    'bcc': bcc,
                    'subject': 'Τεστ test',
                    'body': 'This is a τεστ on utf8'}
            with self.assertRaises(ValueError):
                message = Message(**info)

    def testValid(self):
        info = {'sender': 'test@valid.com',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8'}
        message = Message(**info)
        self.assertTrue(message.subject == u'Τεστ test')
        self.assertTrue(message.body == u'This is a τεστ on utf8')

    def testInvalidContent(self):
        info = {'sender': '',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8',
                'content': 'Texd'}
        with self.assertRaises(ValueError):
            message = Message(**info)

    def testHtmlContent(self):
        info = {'sender': 'test@test.com',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8',
                'content': 'html'}
        message = Message(**info)
        self.assertTrue(message.subject == u'Τεστ test')
        self.assertTrue(message.body == u'This is a τεστ on utf8')

    def testTextContent(self):
        info = {'sender': 'test@test.com',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8',
                'content': 'text'}
        message = Message(**info)
        self.assertTrue(message.subject == u'Τεστ test')
        self.assertTrue(message.body == u'This is a τεστ on utf8')

    def testRecipients(self):
        info = {'sender': 'test@test.com',
                'recipients': 'whatever@gmail.com',
                'cc': ['somebody@gmail.com'],
                'bcc': 'more@gmail.com,andmore@gmail.com',
                'subject': 'Τεστ test',
                'body': 'This is a τεστ on utf8',
                'content': 'html'}
        message = Message(**info)
        self.assertTrue(message.recipients == ['whatever@gmail.com',
                                               'somebody@gmail.com',
                                               'more@gmail.com',
                                               'andmore@gmail.com'])

    def testSmtpInstance(self):
        info = {'smtp_address': 'smtp.test.com',
                'username': 'hacker',
                'password': 'whatever',
                'ssl': False,
                'tls': True,
                'port': 587}
        smtp = SmtpServer(**info)
        self.assertTrue(smtp.address == 'smtp.test.com')
        self.assertTrue(smtp.username == 'hacker')
        self.assertTrue(smtp.password == 'whatever')
        self.assertFalse(smtp.ssl)
        self.assertTrue(smtp.tls)
        self.assertTrue(smtp.port == 587)
        self.assertFalse(smtp.connected)

    def tearDown(self):
        """
        Test tear down

        This is where you should tear down what you've setup in setUp before. This method is called after every test.
        """
        pass


emaillib/_version.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: _version.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Manages the version of the package.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

import os

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''16-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".

VERSION_FILE_PATH = os.path.abspath(
    os.path.join(
        os.path.dirname(__file__),
        '..',
        '.VERSION'
    )
)

LOCAL_VERSION_FILE_PATH = os.path.abspath(
    os.path.join(
        os.path.dirname(__file__),
        '.VERSION'
    )
)

try:
    with open(VERSION_FILE_PATH) as f:
        __version__ = f.read()
except IOError:
    with open(LOCAL_VERSION_FILE_PATH) as f:
        __version__ = f.read()


emaillib/emaillibexceptions.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: emaillibexceptions.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Custom exception code for emaillib.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''16-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


emaillib/emaillib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# File: emaillib.py
#
# Copyright 2017 Costas Tyfoxylos
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
#  of this software and associated documentation files (the "Software"), to
#  deal in the Software without restriction, including without limitation the
#  rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
#  sell copies of the Software, and to permit persons to whom the Software is
#  furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
#

"""
Main code for emaillib.

.. _Google Python Style Guide:
   http://google.github.io/styleguide/pyguide.html

"""

import logging
import os
import re
import smtplib
from email.mime.application import MIMEApplication
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.utils import COMMASPACE, formatdate, parseaddr

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = '''google'''
__date__ = '''16-09-2017'''
__copyright__ = '''Copyright 2017, Costas Tyfoxylos'''
__credits__ = ["Costas Tyfoxylos"]
__license__ = '''MIT'''
__maintainer__ = '''Costas Tyfoxylos'''
__email__ = '''<costas.tyf@gmail.com>'''
__status__ = '''Development'''  # "Prototype", "Development", "Production".


# This is the main prefix used for logging
LOGGER_BASENAME = '''emaillib'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())


class SmtpServer:  # pylint: disable=too-many-instance-attributes
    """A simple wrapper around build in smtplib capabilities."""

    def __init__(self,  # pylint: disable=too-many-arguments
                 smtp_address,
                 username=None,
                 password=None,
                 tls=True,
                 ssl=False,
                 port=None):
        self._smtp = None
        self._address = smtp_address
        self._tls = tls
        self._ssl = ssl
        self._port = port
        self._username = username
        self._password = password
        self._connected = False
        self._logger = logging.getLogger(__name__)

    @property
    def tls(self):
        """The setting of tls upon instantiation."""
        return self._tls

    @property
    def ssl(self):
        """The setting of ssl upon instantiation."""
        return self._ssl

    @property
    def username(self):
        """The username upon instantiation."""
        return self._username

    @property
    def password(self):
        """The password upon instantiation."""
        return self._password

    @property
    def address(self):
        """The smtp server address upon instantiation."""
        return self._address

    @property
    def port(self):
        """The smtp server port upon instantiation."""
        return self._port

    @property
    def connected(self):
        """The status of connection to the smtp server."""
        return self._connected

    def connect(self):
        """Initializes a connection to the smtp server.

        :return: True on success, False otherwise
        """
        connection_method = 'SMTP_SSL' if self.ssl else 'SMTP'
        self._logger.debug('Trying to connect via {}'.format(connection_method))
        smtp = getattr(smtplib, connection_method)
        if self.port:
            self._smtp = smtp(self.address, self.port)
        else:
            self._smtp = smtp(self.address)
        self._smtp.ehlo()
        if self.tls:
            self._smtp.starttls()
            self._smtp.ehlo()
        self._logger.info('Got smtp connection')
        if self.username and self.password:
            self._logger.info('Logging in')
            self._smtp.login(self.username, self.password)
        self._connected = True

    def send(self,  # pylint: disable=too-many-arguments, invalid-name
             sender,
             recipients,
             cc=None,
             bcc=None,
             subject='',
             body='',
             attachments=None,
             content='text'):
        """Sends the email.

        :param sender: The server of the message
        :param recipients: The recipients (To:) of the message
        :param cc: The CC recipients of the message
        :param bcc: The BCC recipients of the message
        :param subject: The subject of the message
        :param body: The body of the message
        :param attachments: The attachments of the message
        :param content: The type of content the message [text/html]

        :return: True on success, False otherwise
        """
        if not self.connected:
            self._logger.error(('Server not connected, cannot send message, '
                                'please connect() first and disconnect() when '
                                'the connection is not needed any more'))
            return False
        try:
            message = Message(sender,
                              recipients,
                              cc,
                              bcc,
                              subject,
                              body,
                              attachments,
                              content)
            self._smtp.sendmail(message.sender,
                                message.recipients,
                                message.as_string)
            result = True
            self._connected = False
            self._logger.debug('Done')
        except Exception:  # noqa
            self._logger.exception('Something went wrong!')
            result = False
        return result

    def disconnect(self):
        """Disconnects from the remote smtp server.

        :return: True on success, False otherwise
        """
        if not self.connected:
            return True
        try:
            self._smtp.close()
            self._connected = False
            result = True
        except Exception:  # noqa
            self._logger.exception('Something went wrong!')
            result = False
        return result


class EasySender:  # pylint: disable=too-few-public-methods, too-many-arguments, invalid-name
    """A simple wrapper around the SmtpServer object."""

    def __init__(self,
                 smtp_address,
                 username=None,
                 password=None,
                 tls=False,
                 ssl=True,
                 port=None):
        self._server = SmtpServer(smtp_address,
                                  username,
                                  password,
                                  tls,
                                  ssl,
                                  port)

    def send(self,
             sender,
             recipients,
             cc=None,
             bcc=None,
             subject='',
             body='',
             attachments=None,
             content='text'):
        """Sends the email by connecting and disconnecting after the send.

        :param sender: The sender of the message
        :param recipients: The recipients (To:) of the message
        :param cc: The CC recipients of the message
        :param bcc: The BCC recipients of the message
        :param subject: The subject of the message
        :param body: The body of the message
        :param attachments: The attachments of the message
        :param content: The type of content the message [text/html]

        :return: True on success, False otherwise
        """
        self._server.connect()
        self._server.send(sender,
                          recipients,
                          cc,
                          bcc,
                          subject,
                          body,
                          attachments,
                          content)
        self._server.disconnect()
        return True


class Message:  # pylint: disable=too-many-instance-attributes
    """A model of an email message."""

    def __init__(self,  # pylint: disable=too-many-arguments
                 sender,
                 recipients,
                 cc=None,
                 bcc=None,
                 subject='',
                 body='',
                 attachments=None,
                 content='text'):
        self._sender = None
        self._recipients = None
        self._to = []
        self._cc = []
        self._bcc = []
        self._subject = None
        self._body = None
        self._attachments = None
        self._parts = None
        self._content = None
        self._message = None
        self.sender = sender
        self.to = recipients  # pylint: disable=invalid-name
        self.cc = cc  # pylint: disable=invalid-name
        self.bcc = bcc
        self.subject = subject
        self.body = body
        self.attachments = attachments
        self.content = content
        self._setup_message()

    def _setup_message(self):
        """Constructs the actual underlying message with provided values.."""
        if self.content == 'html':
            self._message = MIMEMultipart('alternative')
            part = MIMEText(self.body, 'html', 'UTF-8')
        else:
            self._message = MIMEMultipart()
            part = MIMEText(self.body, 'plain', 'UTF-8')
        self._message.preamble = 'Multipart massage.\n'
        self._message.attach(part)
        self._message['From'] = self.sender
        self._message['To'] = COMMASPACE.join(self.to)
        if self.cc:
            self._message['Cc'] = COMMASPACE.join(self.cc)
        self._message['Date'] = formatdate(localtime=True)
        self._message['Subject'] = self.subject
        for part in self._parts:
            self._message.attach(part)

    @staticmethod
    def _validate_simple(email):
        """Does a simple validation of an email by matching it to a regex.

        :param email: The email to check
        :return: The valid Email address

        :raises: ValueError if value is not a valid email
        """
        _, address = parseaddr(email)
        if not re.match(r'[^@]+@[^@]+\.[^@]+', address):
            raise ValueError('Invalid email :{email}'.format(email=email))
        return address

    @staticmethod
    def _comma_delimited_to_list(value):
        if not isinstance(value, (list, tuple)):
            return [entry.strip() for entry in value.split(',')]
        return value

    @property
    def sender(self):
        """The email address of the sender."""
        return self._sender

    @sender.setter
    def sender(self, value):
        if not value:
            raise ValueError('Sender cannot be empty.')
        self._sender = self._validate_simple(value)

    @property
    def recipients(self):
        """A list of all recipients of the message."""
        return self.to + self.cc + self.bcc

    def _get_recipients(self, value):
        recipients = self._comma_delimited_to_list(value) if value else []
        return [self._validate_simple(recipient) for recipient in recipients]

    @property
    def to(self):  # pylint: disable=invalid-name
        """The main (to) recipients of the message."""
        return self._to

    @to.setter
    def to(self, value):  # pylint: disable=invalid-name
        if not value:
            raise ValueError('Recipients cannot be empty.')
        self._to = self._get_recipients(value)

    @property
    def cc(self):  # pylint: disable=invalid-name
        """The cc recipients of the message."""
        return self._cc

    @cc.setter
    def cc(self, value):  # pylint: disable=invalid-name
        self._cc = self._get_recipients(value)

    @property
    def bcc(self):
        """The bcc recipients of the message."""
        return self._bcc

    @bcc.setter
    def bcc(self, value):
        self._bcc = self._get_recipients(value)

    @property
    def subject(self):
        """The subject of the message."""
        return self._subject

    @subject.setter
    def subject(self, value):
        self._subject = value

    @property
    def body(self):
        """The body of the message."""
        return self._body

    @body.setter
    def body(self, value):
        self._body = value

    @property
    def content(self):
        """The type of content of the message."""
        return self._content

    @content.setter
    def content(self, value):
        if value.lower() not in ['text', 'html']:
            raise ValueError(('Invalid content type :{content}.'
                              'Allowed ["text", "html"]').format(content=value))
        self._content = value.lower()

    @property
    def attachments(self):
        """A list of attachment names of the message."""
        return self._attachments

    @attachments.setter
    def attachments(self, entries):
        values = []
        self._parts = []
        attachments = self._comma_delimited_to_list(entries) if entries else []
        for attachment in attachments:
            part = MIMEApplication(open(os.path.expanduser(attachment),
                                        "rb").read())
            part.add_header('Content-Disposition',
                            'attachment',
                            filename=os.path.basename(attachment))
            self._parts.append(part)
            values.append(attachment)
        self._attachments = values

    @property
    def as_string(self):
        """The string representation of the message."""
        return self._message.as_string()

    def __str__(self):
        """The string representation of the message."""
        return self.as_string
